### PR TITLE
Adds basic support for pin and unpin operations

### PR DESCRIPTION
**Pull request by saresend**
## What
Should address #11, by adding some basic pin / unpin operations to the dag

## Todo
- [x] Add support for querying whether a block is pinned
- [ ] Add Tests once design is finished

### CODE REVIEW

**Comment by aphelionz** [comment]
This is really exciting, thank you @saresend. If this works for @koivunej then it'd be great to see tests, and then if you want to go for the gold, add the http endpoints! 🎉

**Comment by dvc94ch** [comment]
I guess remove should check if the block is pinned.

**Comment by saresend** [comment]
Alright, a couple of changes: 
* First, pinning is defined by the existence of a pair in the Pin hashmap, rather than a boolean flag of 1 or 0. This way we don't have two `False` states, in the form of `Null` or `Zero`, and instead consolidate all into the null case.

* It should now also error when trying to unpin an unpinned block, rather than just silently performing a no-op.

**Comment by aphelionz** [comment]
@saresend Here's the overall spec, including what the conformance tests expect for the PIN apis. I hope this is helpful. Let me know if you have any questions

https://github.com/ipfs-rust/ipfs-rust-conformance/issues/2

**Comment by saresend** [comment]
I was going to add some tests at the very least. I might leave the actual `http` endpoints to a separate PR however. If you have some code points for how you write tests for this project, I'd be happy to add them.

**Comment by aphelionz** [comment]
@saresend @dvc94ch Tests and doc would be important. You can add simple module tests at the end of the respective files, e.g. `dag.rs` and `repo/mod.rs`. It would also be huge if you could do the endpoints. There's more examples of testing in that crate as well.

Thank you so much again for this.

**Comment by saresend** [comment]
No problem, I'll open up a new PR for the http endpoints, just to keep concerns separate!

**Comment by dvc94ch** [comment]
It's also not clear why you're exposing it on the dag, why unpin should error if a block already is unpinned, as stuff may share blocks or how this works as you can still remove a block even if it's pinned.

**Comment by saresend** [comment]
Totally fair, I'm still ramping up on IPFS and don't have great context on how you envision this being architected or how concerns are being separated. If you have suggestions on better ways to expose the API I'm all ears. My understanding of pinning is that it protects against garbage collection, and I wasn't able to find whether garbage collection is already being done in `rust-ipfs`. My thought was that garbage collection could leverage the pinning query logic when removing nodes whenever it does get built

**Comment by aphelionz** [comment]
As far as the error goes, it should simply follow what js-ipfs or go-ipfs does. So if they error when trying to unpin and already unpinned block, then we should too. Otherwise not.

**Comment by dvc94ch** [comment]
I don't think the dag should expose pin/unpin, and if it did then the semantics would probably not be to pin the root of a path as a path can point to other blocks, so you'd want to pin the last block it refers to.

Yes pin is for garbage collection, which is not currently implemented. What currently is implemented is manually removing blocks, which shouldn't happen if it's pinned.

Since blocks can be shared, pinning should probably increment a counter and unpin should decrement it, instead of having a boolean.

Garbage collection can be implemented by iterating through all blocks, checking if the pin count is zero and if it is remove the block.

**Comment by saresend** [comment]
Where would you recommend pinning be implemented?

File: src/lib.rs [hunk 1]
```diff
@@ -312,6 +312,21 @@ impl<Types: IpfsTypes> Ipfs<Types> {
         Ok(self.repo.remove_block(cid).await?)
     }
 
+    /// Pins a given Cid
+    pub async fn pin_block(&self, path: IpfsPath) -> Result<(), Error> {
```
Comment (dvc94ch) [inline src/lib.rs]:
```suggestion
    pub async fn pin_block(&self, cid: &Cid) -> Result<(), Error> {
        self.repo.pin_block(cid).await
    }
```

**Comment by aphelionz** [comment]
I might suggest we focus on tests and the HTTP endpoints here, so that at least there's a foundation for architecture changes + refactors. Always room for improvement, especially if we keep getting awesome community contributions.

**Comment by dvc94ch** [comment]
Do you have anything to add in regards to my comment? Did I not explain in which ways the implementation suggested here is flawed clearly enough? Conformance tests are not a way to avoid using your brain, as tests are just as buggy as code. Besides tests do not ensure correctness, they sometimes show the presence of bugs.

**Comment by dvc94ch** [comment]
Let me try again, I'll use letters to denote blocks in the following diagram:

```txt
a -> b <- c
```

and these two concurrent processes running in parallel

```txt
process 1:
    put(b)
    pin(b)
    put(a)
    pin(a)
    ...
    unpin(a)
    unpin(b)
    gc()
```
```txt
process 2:
    put(b)
    pin(b)
    put(c)
    pin(c)

    get(b)???
```

**Comment by dvc94ch** [comment]
Also put should probably pin automatically to prevent a race, but that's something to be discussed, if someone can supply a usecase where you don't want to put_and_pin

**Comment by dvc94ch** [comment]
I approved the PR and would have merged it if it didn't have conflicts with master. But I agree with @aphelionz, this features flaws should be extensively documented and tested.

**Comment by saresend** [comment]
Before the next couple changes, I just want to clarify what you'd want for this PR. So far, the changes requested are:
* Move to use `&Cid`, rather than `ipfsPath`
* Automatically Pin on Put actions
* Prevent removal based on whether a block is pinned
* Add tests and documentation

Is there anything I'm missing here?

**Comment by dvc94ch** [comment]
If we didn't have the conformance tests I'd suggest removing remove altogether. Also don't use a bool as shown in the example above it can lead to issues. But mainly I'm a bit annoyed that my approval on pr's has been purchased for a mere 2000€, which has nothing to do with your pr. Thanks for contributing...

**Comment by saresend** [comment]
@dvc94ch @aphelionz, alright this should include a more `RC` like logic for pinning, removes it from the Dag API, and finally includes a test to verify behavior. It should also now be rebased so if it all checks out feel free to merge.

**Comment by dvc94ch** [comment]
Looks good! So I think this is much better, thanks. Can you rebase again, looks like it's not mergeable.

**Comment by saresend** [comment]
Alright that should be merged!

**Comment by koivunej** [comment]
Interesting the same test is flaky on windows ... Again. Need to investigate why.

**Comment by koivunej** [comment]
The test failure will be fixed with #124 but I think this is now good to merge?

**Comment by aphelionz** [comment]
@saresend Please add yourself to https://github.com/ipfs-rust/welcome/issues/1 and also let us know if you plan on doing the endpoints as well

**Comment by saresend** [comment]
Yep! I plan to have a PR up in the next week or so, classes have started back up again for me so it may take me a bit longer unfortunately

**Comment by aphelionz** [comment]
That's ok, no rush.
